From 5c9043d4db407bba20746d0d83928f38d7fe9820 Mon Sep 17 00:00:00 2001
From: Justin Ku <chunku@qti.qualcomm.com>
Date: Wed, 11 Feb 2026 18:47:20 -0800
Subject: [PATCH] [linux net] Add hooks infrastructure for PON platform

These features default disable
The changes introduce a hook mechanism (pon_iftrap) in the network stack
(receive and transmit paths) specifically for the PON (Passive Optical Network) platform
to intercept and process related packets

Change-Id: Ifce747dfb758f527c4d8704fb099cb1f110f6375
Signed-off-by: Justin Ku <chunku@qti.qualcomm.com>
---
 include/linux/netdevice.h | 16 ++++++++
 include/linux/skbuff.h    |  5 +++
 net/core/dev.c            | 81 +++++++++++++++++++++++++++++++++++++++
 3 files changed, 102 insertions(+)

diff --git a/include/linux/netdevice.h b/include/linux/netdevice.h
index 6f673c497620..599b247c0f3e 100644
--- a/include/linux/netdevice.h
+++ b/include/linux/netdevice.h
@@ -5527,4 +5527,20 @@ int netdev_hw_offload_ops_unregister(struct net_device *dev, struct netdev_hw_of
 		atomic_long_add((VAL), &(DEV)->stats.__##FIELD)
 #define DEV_STATS_READ(DEV, FIELD) atomic_long_read(&(DEV)->stats.__##FIELD)
 
+/* This structure is passed to the PON trap handler to provide context
+ * about which transmit path is being used, allowing the handler to make
+ * path-specific decisions.
+ */
+
+#ifdef CONFIG_IPQ_PON
+#define __DEV_QUEUE_XMIT 0
+#define DEV_FAST_XMIT 1
+#define DEV_FAST_XMIT_VP 2
+#define DEV_FAST_XMIT_QDISC 3
+
+struct pon_iftrap_devq_uni_tx_meta {
+	unsigned char calling_id;
+};
+#endif
+
 #endif	/* _LINUX_NETDEVICE_H */
diff --git a/include/linux/skbuff.h b/include/linux/skbuff.h
index 842d4b458bf7..91b6f4e87ee1 100644
--- a/include/linux/skbuff.h
+++ b/include/linux/skbuff.h
@@ -1018,6 +1018,11 @@ struct sk_buff {
 	__u8			int_pri:4;
 	/* Priority info for hardware qdiscs */
 
+#ifdef CONFIG_IPQ_PON
+	/* packets processed by pon_iftrap_......() */
+	__u8			pon_softpath:2;
+#endif
+
 #if defined(CONFIG_NET_SCHED) || defined(CONFIG_NET_XGRESS)
 	__u16			tc_index;	/* traffic control index */
 #endif
diff --git a/net/core/dev.c b/net/core/dev.c
index 839453085a14..87a19469f0e0 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -2125,6 +2125,18 @@ static int call_netdevice_notifiers_mtu(unsigned long val,
 bool fast_tc_filter;
 EXPORT_SYMBOL_GPL(fast_tc_filter);
 
+#ifdef CONFIG_IPQ_PON
+/* By default PON Trap Filter will be disabled. */
+bool pon_iftrap_enable;
+EXPORT_SYMBOL_GPL(pon_iftrap_enable);
+
+int (*pon_iftrap_recv)(struct sk_buff *skb) __rcu __read_mostly;
+EXPORT_SYMBOL_GPL(pon_iftrap_recv);
+
+int (*pon_iftrap_xmit)(void *app_data, struct sk_buff *skb) __rcu __read_mostly;
+EXPORT_SYMBOL_GPL(pon_iftrap_xmit);
+#endif
+
 #ifdef CONFIG_NET_INGRESS
 static DEFINE_STATIC_KEY_FALSE(ingress_needed_key);
 
@@ -4649,6 +4661,31 @@ struct netdev_queue *netdev_core_pick_tx(struct net_device *dev,
 	return netdev_get_tx_queue(dev, queue_index);
 }
 
+#ifdef CONFIG_IPQ_PON
+/*	When the external module is loaded, it sets the values of
+ *	pon_iftrap_xmit and pon_iftrap_recv during the module initialization phase,
+ *	and only then sets pon_iftrap_enable to 1.
+ */
+static inline bool pon_iftrap_process_tx(struct sk_buff *skb, unsigned char calling_id)
+{
+	int (*pon_uni_xmit)(void *app_data, struct sk_buff *skb);
+	struct pon_iftrap_devq_uni_tx_meta meta;
+
+	if (pon_iftrap_enable) {
+		/* Caller must hold RCU read lock */
+		pon_uni_xmit = rcu_dereference_bh(pon_iftrap_xmit);
+		if (pon_uni_xmit) {
+			meta.calling_id = calling_id;
+			if (pon_uni_xmit(&meta, skb)) {
+				/* Return true if packet was consumed or dropped */
+				return true;
+			}
+		}
+	}
+	return false; /* Return false if packet need to continue */
+}
+#endif
+
 /**
  *	dev_fast_xmit_vp - fast xmit the skb to a PPE virtual port
  *	@skb:buffer to transmit
@@ -4672,6 +4709,14 @@ bool dev_fast_xmit_vp(struct sk_buff *skb,
 	}
 
 	rcu_read_lock_bh();
+
+#ifdef CONFIG_IPQ_PON
+	if (pon_iftrap_process_tx(skb, DEV_FAST_XMIT_VP)) {
+		rcu_read_unlock_bh();
+		return true;
+	}
+#endif
+
 	cpu = smp_processor_id();
 
 	/*
@@ -4743,6 +4788,13 @@ bool dev_fast_xmit_qdisc(struct sk_buff *skb, struct net_device *top_qdisc_dev,
 	 */
 	rcu_read_lock_bh();
 
+#ifdef CONFIG_IPQ_PON
+	if (pon_iftrap_process_tx(skb, DEV_FAST_XMIT_QDISC)) {
+		rcu_read_unlock_bh();
+		return true;
+	}
+#endif
+
 	txq = netdev_core_pick_tx(top_qdisc_dev, skb, NULL);
 	q = rcu_dereference_bh(txq->qdisc);
 	if (unlikely(!q->enqueue)) {
@@ -4815,6 +4867,14 @@ bool dev_fast_xmit(struct sk_buff *skb,
 	}
 
 	rcu_read_lock_bh();
+
+#ifdef CONFIG_IPQ_PON
+	if (pon_iftrap_process_tx(skb, DEV_FAST_XMIT)) {
+		rcu_read_unlock_bh();
+		return true;
+	}
+#endif
+
 	cpu = smp_processor_id();
 
 	/* If device don't need the dst, release it now, otherwise make sure
@@ -4906,6 +4966,13 @@ int __dev_queue_xmit(struct sk_buff *skb, struct net_device *sb_dev)
 
 	skb_update_prio(skb);
 
+#ifdef CONFIG_IPQ_PON
+	if (pon_iftrap_process_tx(skb, __DEV_QUEUE_XMIT)) {
+		rc = NETDEV_TX_OK;
+		goto out;
+	}
+#endif
+
 	qdisc_pkt_len_init(skb);
 	tcx_set_ingress(skb, false);
 #ifdef CONFIG_NET_EGRESS
@@ -6020,6 +6087,20 @@ static int __netif_receive_skb_core(struct sk_buff **pskb, bool pfmemalloc,
 		}
 	}
 
+#ifdef CONFIG_IPQ_PON
+	if (pon_iftrap_enable) {
+		int (*pon_uni_recv)(struct sk_buff *skb);
+
+		pon_uni_recv = rcu_dereference(pon_iftrap_recv);
+		if (pon_uni_recv) {
+			if (pon_uni_recv(skb)) { //consumed or dropped
+				ret = NET_RX_DROP;
+				goto out;
+			}
+		}
+	}
+#endif
+
 	if (likely(!fast_tc_filter)) {
 		fast_recv = rcu_dereference(athrs_fast_nat_recv);
 		if (fast_recv) {
-- 
2.34.1

